### PR TITLE
Refresh expired websocket sessions

### PR DESCRIPTION
**Pull request by richardm-stripe**
### Reviewers
r? @ob-stripe 
cc @tomer-stripe 

 ### Summary

- `stripe listen` and `stripe logs tail` now detect when a CLI session is expired, and attempt to create a new one and open a websocket connection with it up to three times, this should fix
https://github.com/stripe/stripe-cli/issues/80

- Change use of `context.Context` to be more idiomatic. From [the docs](https://golang.org/pkg/context/)
> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx:

- Factored `withSIGTERMCancel` out of `proxy.Run` and `tailer.Run`. 

### Testing

- Hardcoded devbox stripecli to return the expired error, verified the command terminates after 3 attempts. Verified that the command can successfully reconnect when I un-hardcode the expiry halfway through.
- Against production, I disconnect wi-fi for 15 minutes and verify the command successfully reconnects. You can see the logs here: https://gist.github.com/richardm-stripe/356cf5fb7e6c802efd2adcc764adc157

### CODE REVIEW

File: pkg/logtailing/tailer.go [hunk 1]
```diff
@@ -107,60 +105,84 @@ func New(cfg *Config) *Tailer {
...
+		go t.webSocketClient.Run(ctx)
+		nAttempts++
+
+		select {
+		case <-ctx.Done():
+			ansi.StopSpinner(s, "", t.cfg.Log.Out)
+			t.cfg.Log.Fatalf("Aborting")
+		case <-t.webSocketClient.NotifyExpired:
+			if nAttempts < maxConnectAttempts {
+				s = ansi.StartSpinner("Session expired, reconnecting...", t.cfg.Log.Out)
```
Comment (ob-stripe) [inline pkg/logtailing/tailer.go:180]: Do we need to stop the old spinner first? Or does starting a new spinner automatically stops the existing one?

Comment (richardm-stripe) [reply pkg/logtailing/tailer.go]: Good call. I had not been testing with color on. It was a jittery mess. I fixed it in the latest.

File: pkg/logtailing/tailer.go [hunk 2]
```diff
@@ -107,60 +105,84 @@ func New(cfg *Config) *Tailer {
...
 
 	go func() {
+		<-interruptCh
+		onCancel()
+		cancel()
+	}()
+	return ctx
+}
+
+const maxConnectAttempts = 3
```
[the remark this answers is not in the document]
Comment (richardm-stripe) [reply pkg/logtailing/tailer.go:123]: I doubt it? If refreshing doesn't work on the first attempt, it's very unlikely to work on subsequent attempts. So if bumping this value makes any material difference it means something else is wrong.

**Comment by richardm-stripe** [comment]
ptal @ob-stripe 

Added a test, fixed the spinner

**Pull request merged**